src/cdomains/congruenceClosure.ml

```diff
@@ -0,0 +1,1581 @@
+          | TInt _ -> "Int"
+          | TArray _ -> "Arr"
+          | TVoid _ -> "Voi"
+          | TFloat (_, _)-> "Flo"
+          | TComp (_, _) -> "TCo"
+          | TFun (_, _, _, _)|TNamed (_, _)|TEnum (_, _)|TBuiltin_va_list _ -> "?"
+        )^ Z.to_string (get_size_in_bits typ) ^ "]"
+    with
+    | UnsupportedCilExpression _ -> "[?]"
+
+  let rec show : t -> string = function
+    | Addr v -> "&" ^ Var.show v
+    | Deref (Addr v, z, exp) when Z.equal z Z.zero -> Var.show v ^ show_type exp
+    | Deref (t, z, exp) when Z.equal z Z.zero -> "*" ^ show t^ show_type exp
+    | Deref (t, z, exp) -> "*(" ^ Z.to_string z ^ "+" ^ show t ^ ")"^ show_type exp
+
+  (** Returns true if the first parameter is a subterm of the second one. *)
+  let rec is_subterm st term = equal st term || match term with
+    | Deref (t, _, _) -> is_subterm st t
+    | _ -> false
+
+  let rec get_var = function
+    | Addr v -> v
+    | Deref (t, _, _) -> get_var t
+
+  (** Returns true if the second parameter contains one of the variables defined in the list "variables". *)
+  let rec contains_variable variables term = List.mem_cmp Var.compare (get_var term) variables
+
+  let term_of_varinfo vinfo =
+    Deref (Addr vinfo, Z.zero, Lval (Var vinfo, NoOffset))
+
+  let eval_int (ask:Queries.ask) exp =
+    match Cilfacade.get_ikind_exp exp with
+    | exception Invalid_argument _ -> raise (UnsupportedCilExpression "non-constant value")
+    | ikind ->
+      begin match ask.f (Queries.EvalInt exp) with
+        | `Lifted i ->
+          begin match IntDomain.IntDomTuple.to_int @@ IntDomain.IntDomTuple.cast_to ikind i
+            with
+            | Some i -> i
+            | None -> raise (UnsupportedCilExpression "non-constant value")
+          end
+        | _ -> raise (UnsupportedCilExpression "non-constant value")
+      end
+
+  let eval_int_opt (ask:Queries.ask) exp =
+    match eval_int ask exp with
+    | i -> Some i
+    | exception (UnsupportedCilExpression _) -> None
+
+  (*returns Some type for a pointer to a type
+    and None if the result is not a pointer*)
+  let rec type_of_element typ =
+    match typ with
+    | TArray (typ, _, _) -> type_of_element typ
+    | TPtr (typ, _) -> Some typ
+    | _ -> None
+
+  (** Returns the size of the type. If typ is a pointer, it returns the
+      size of the elements it points to. If typ is an array, it returns the size of the
+      elements of the array (even if it is a multidimensional array. Therefore get_element_size_in_bits int\[]\[]\[] = sizeof(int)). *)
+  let rec get_element_size_in_bits typ =
+    match type_of_element typ with
+    | Some typ -> get_size_in_bits typ
+    | None -> Z.one
+
+  let is_array_type = function
+    | TArray _ -> true
+    | _ -> false
+
+  let is_struct_type = function
+    | TComp _ -> true
+    | _ -> false
+
+  let is_struct_ptr_type = function
+    | TPtr(TComp _,_) -> true
+    | _ -> false
+
+  let is_ptr_type = function
+    | TPtr _ -> true
+    | _ -> false
```

Comment:
Before calling these, you might want to unroll first to get rid of any `TNamed` (typedefs) that may surround them